generic/tclProc.c

```diff
@@ -1513,6 +1536,337 @@ InitArgsAndLocals(
+                            }
+                        }
+                    }
+                }
+            }
+
+            if (!found) {
+                /* we did not find the key value in the arguments list, add the default value. */
+                if (!varPtr->defValuePtr) {
+                    /* No default value, that's an error by the caller. */
+                    goto error_args;
+                }
+                Tcl_IncrRefCount(varPtr->defValuePtr);
+                Tcl_ListObjAppendElement(interp, copyPtr, varPtr->defValuePtr);
+            }
+
+            /* move to next parameter */
+            varPtr = varPtr->nextPtr;
+        }
+    }
+
+    {
+        /* Copy the positional parameters. */
+        int len = 0;
+
+        /* The named parameters or default values are all on the new argument list. Add the marker */
+        Tcl_Obj *marker = Tcl_NewStringObj("--", 2);
+        Tcl_IncrRefCount(marker);
+        Tcl_ListObjAppendElement(interp, copyPtr, marker);
+
+        /* Skip the -- parameter if present in calling arguments */
+        const char *keyString = TclGetString(framePtr->objv[srcIndex]);
+        if (keyString && keyString[0] == '-' && keyString[1] == '-' && !keyString[2]) {
+            srcIndex++;
+        }
+
+        while (srcIndex < framePtr->objc) {
+            Tcl_ListObjAppendElement(NULL, copyPtr, framePtr->objv[srcIndex]);
+            Tcl_IncrRefCount(framePtr->objv[srcIndex]);
+            srcIndex++;
+        }
+
+        /* Check that there is a value for every positional parameter */
+        Tcl_ListObjLength(NULL, copyPtr, &len);
+        if (len < framePtr->procPtr->numArgs + 1 - (varArgs ? 1 : 0)) {
+            /* There are position parameters we need to fill in if they have defaults. */
+            CompiledLocal *varPtr = framePtr->procPtr->firstLocalPtr;
+            int index = 0;
+            /* skip the parameters we have already added */
+            while (varPtr && --len > 0) {
+                varPtr = varPtr->nextPtr;
+                index++;
+            }
+            /* Copy any remaining parameters with defaults */
+            while (varPtr) {
+                Tcl_Obj *defValuePtr = varPtr->defValuePtr;
+                if (defValuePtr) {
+                    Tcl_ListObjAppendElement(NULL, copyPtr, defValuePtr);
+                    Tcl_IncrRefCount(defValuePtr);
+                }
+                /* skip the final argument if varargs */
+                if (index != framePtr->procPtr->numArgs - 1 || !varArgs) {
+                    varPtr = varPtr->nextPtr;
+                } else {
+                    varPtr = NULL;
+                }
+                index++;
+            }
+        }
+    }
+
+    {
+        /* Check that all of the parameters have a value. */
+        Tcl_Obj ** newObjv;
+        int newObjc;
+
+        /* Get the pointer to the object array */
+        Tcl_ListObjGetElements(NULL, copyPtr, &newObjc, &newObjv);
```

Comment:
Tcl_ListObjGetElements can fail and return TCL_ERROR.  Perhaps that's not possible here, though.